purchase_order_line_schedule/models/purchase_order_line.py

```diff
@@ -0,0 +1,101 @@
+    )
+
+    @api.depends("product_qty", "schedule_line_ids", "schedule_line_ids.product_qty")
+    def _compute_qty_to_schedule(self):
+        for line in self:
+            line.qty_to_schedule = line.product_qty - sum(
+                line.schedule_line_ids.mapped("product_qty")
+            )
+
+    @api.depends("schedule_line_ids")
+    def _compute_has_schedule_lines(self):
+        for line in self:
+            line.has_schedule_lines = line.schedule_line_ids and True or False
+
+    @api.depends("schedule_line_ids", "schedule_line_ids.qty_received")
+    def _compute_qty_received(self):
+        super(PurchaseOrderLine, self)._compute_qty_received()
+        for line in self:
+            if line.schedule_line_ids:
+                line.qty_received = sum(
+                    sl.qty_received for sl in line.schedule_line_ids
+                )
+
+    @api.onchange("qty_received")
+    def _inverse_qty_received(self):
+        for line in self:
+            if line.qty_received_method == "manual":
+                qty_to_allocate = line.qty_received
+                for sl in line.schedule_line_ids.sorted(lambda x: x.date_planned):
+                    sl.qty_received = min(sl.product_qty, qty_to_allocate)
+                    qty_to_allocate -= sl.qty_received_manual
+            else:
+                super(PurchaseOrderLine, line)._inverse_qty_received()
+
+    def action_open_schedule_lines(self):
+        line_ids = []
+        for line in self:
+            line_ids += line.schedule_line_ids.ids
+
+        domain = [("id", "in", line_ids)]
+        context = dict(self.env.context)
+        context.update(default_order_line_id=self.id)
+
+        return {
+            "name": _("Schedule Lines"),
+            "type": "ir.actions.act_window",
+            "res_model": "purchase.order.line.schedule",
+            "view_mode": "tree,form",
+            "target": "current",
+            "domain": domain,
+            "context": context,
+        }
+
+    def _update_schedule_lines(self):
+        for rec in self:
+            sls = rec.schedule_line_ids
+            if not sls:
+                vals = {
+                    "date_planned": rec.date_planned,
+                    "product_qty": rec.product_qty,
+                }
+                rec.schedule_line_ids = [(0, 0, vals)]
+            else:
+                qty = sum(rec.schedule_line_ids.mapped("product_qty"))
+                # Update the last scheduled date of the schedule lines
+                # And update the quantity of schedule lines so as to match
+                # that of the purchase order line.
+                for sl in sls.sorted(lambda x: x.date_planned, reverse=True):
+                    if qty != rec.product_qty:
+                        sl.product_qty += rec.product_qty - qty
+                    if sl.date_planned != rec.date_planned:
+                        sl.date_planned = rec.date_planned
+                    break
```

Comment:
This auto spread feature might be confusing to end-users. Maybe it's better to only update scheduled lines if there is just one scheduled line. If not the user should do it manually?

Reply:
This is necessary to ensure compatibility with existing modules / flows. If a user changes the quantity to order by default it allocates the extra quantity to the most recent schedule line. The user can always to go to the schedule lines and adjust.

Reply:
Maybe we need a constraint in that case on product_qty on SL? The qty should be qty of related PO line. Right now a loop is created if you schedule more or less, the quantities are not saved.

Reply:
Fixed after adding the wizard.